File: backend/src/main/java/com/example/template/services/VentaService.java
package com.example.template.services;

import com.example.template.dtos.*;
import com.example.template.exceptions.VentaDontFoundException;
import com.example.template.model.*;
import com.example.template.repositories.*;
import com.example.template.services.IVentaService;
import com.example.template.mappers.*;
import lombok.AllArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

@Service
@AllArgsConstructor
public class VentaService implements IVentaService {

    protected VentaRepository VentaRepository;
    protected LibroRepository libroRepository;
    protected AutorRepository autorRepository;
    protected MapperDetalleVenta mapperDetalleVenta;
    protected MapperVenta mapperVenta;


        @Override
    public void create(VentaCreateRequestDTO ventaCreateRequestDTO) {
       Venta venta;
       venta = mapperVenta.CreateDtoToVenta(ventaCreateRequestDTO);
       
	   for(DetalleVentaCreateRequestDTO detalleVentaDto : ventaCreateRequestDTO.getDetalles()){
	       DetalleVenta detalleVenta = mapperDetalleVenta.toDetalleVenta(detalleVentaDto);
	
		   if(detalleVentaDto.getLibroId() != 0) {
		       Optional<Libro> optionalLibro = libroRepository.findById(detalleVentaDto.getLibroId());
		
		       if(optionalLibro.isPresent()){
		       detalleVenta.setLibro(optionalLibro.get());
		       }
		    }
	     venta.getDetalles().add(detalleVenta);
	   }
       VentaRepository.save(venta);
    }
    @Override
    public VentaGetByIdResponseDTO getById(long id) throws VentaDontFoundException {
        Optional<Venta> optional = VentaRepository.findById(id);
        if (optional.isEmpty()){
            throw new VentaDontFoundException(id);
        }
        Venta venta = optional.get();


        VentaGetByIdResponseDTO ventaGetByIdResponseDTO = mapperVenta.toVentaGetByIdResponseDTO(venta);

        return ventaGetByIdResponseDTO;
    }

    @Override
    public List<VentaResponseDTO> getAll() {
        List<Venta> VentaList;
        List<VentaResponseDTO> VentaResponseDTOList = new ArrayList<>();

        VentaList = VentaRepository.findAll();

        for (Venta Venta: VentaList) {
                VentaResponseDTO VentaResponseDTO = mapperVenta.toVentaResponseDTO(Venta);
                VentaResponseDTOList.add(VentaResponseDTO);
        }
        return VentaResponseDTOList;
    }
    @Override
    public List<VentaGetAllForAssociationResponseDTO> getAllForAssociation() {
        List<Venta> VentaList;
        List<VentaGetAllForAssociationResponseDTO> VentaGetAllForAssociationResponseDTOList = new ArrayList<>();

        VentaList = VentaRepository.findAll();

        for (Venta Venta: VentaList) {
                VentaGetAllForAssociationResponseDTO VentaGetAllForAssociationResponseDTO = mapperVenta.toVentaGetAllForAssociationResponseDTO(Venta);
                VentaGetAllForAssociationResponseDTOList.add(VentaGetAllForAssociationResponseDTO);
        }
        return VentaGetAllForAssociationResponseDTOList;
    }

    @Override
    public VentaResponseDTO update(long id, VentaRequestDTO VentaRequestDTO) throws VentaDontFoundException {
        Optional<Venta> VentaOptional  = VentaRepository.findById(id);
        if (VentaOptional.isEmpty()){
            throw new VentaDontFoundException(id);
        }
        Venta Venta = VentaOptional.get();

        Venta VentaUpdate = mapperVenta.toVenta(VentaRequestDTO,id);

       VentaUpdate = VentaRepository.save(VentaUpdate);

        VentaResponseDTO VentaResponseDTO = mapperVenta.toVentaResponseDTO(VentaUpdate);

        return VentaResponseDTO;
    }

    @Override
    public boolean delete(long id) throws VentaDontFoundException {
        Optional<Venta> VentaOptional  = VentaRepository.findById(id);

        if (VentaOptional.isEmpty()){
            throw new VentaDontFoundException(id);
        }

        Venta Venta = VentaOptional.get();
        VentaRepository.delete(Venta);

        return true;
    }
}

File: frontend/src/services/Venta/VentaService.ts
import HttpClient from './../HttpClient';
import VentaGetAllResponseDto from '../../dtos/Venta/getAll/VentaGetAllResponseDto';
import VentaGetByIdResponseDto from '../../dtos/Venta/getById/VentaGetByIdResponseDto';
import VentaGetAllForAssociationResponseDto from '../../dtos/Venta/getAllForAssociation/VentaGetAllForAssociationResponseDto';
import VentaCreateRequestDto from '../../dtos/Venta/create/VentaCreateRequestDto';

class VentaService
{
   getAllVenta = async (): Promise<VentaGetAllResponseDto[]> => {
       const response = await HttpClient.get(`api/v1/Venta/`);
       const bodyReponse = await response.json();
       return bodyReponse as VentaGetAllResponseDto[];
   };
   getAllForAssociationVenta = async (): Promise<VentaGetAllForAssociationResponseDto[]> => {
       const response = await HttpClient.get(`api/v1/Venta/getAllForAssociations`);
       const bodyReponse = await response.json();
       return bodyReponse as VentaGetAllForAssociationResponseDto[];
   };

    getById = async (id:number): Promise<VentaGetByIdResponseDto> => {
       const response = await HttpClient.get(`api/v1/Venta/${id}/`);
       const bodyResponse = await response.json();
       return bodyResponse as VentaGetByIdResponseDto;
    }

   postCreateVenta = async (ventaCreateRequestDto : VentaCreateRequestDto): Promise<Response> => {
       return await HttpClient.post(`api/v1/Venta/`, ventaCreateRequestDto);
   };

   deleteVenta = async (id : number): Promise<Response> => {
       return await HttpClient.delete(`api/v1/Venta/${id}`);
   };


}

export default new VentaService()
